h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/predict/UpdateAuxTreeWeightsTask.java

```diff
@@ -0,0 +1,85 @@
+package hex.tree.xgboost.predict;
+
+import biz.k11i.xgboost.Predictor;
+import biz.k11i.xgboost.gbm.GBTree;
+import biz.k11i.xgboost.tree.RegTree;
+import biz.k11i.xgboost.tree.RegTreeNode;
+import hex.DataInfo;
+import hex.tree.xgboost.XGBoostModelInfo;
+import hex.tree.xgboost.XGBoostOutput;
+import water.MRTask;
+import water.fvec.Chunk;
+import water.fvec.NewChunk;
+import water.util.ArrayUtils;
+
+public class UpdateAuxTreeWeightsTask extends MRTask<UpdateAuxTreeWeightsTask> {
+
+    // IN
+    private final Predictor _p;
+    private final DataInfo _di;
+    private final boolean _sparse;
+    // OUT
+    private final double[/*treeId*/][/*leafNodeId*/] _nodeWeights;
+
+    public UpdateAuxTreeWeightsTask(DataInfo di, XGBoostModelInfo modelInfo, XGBoostOutput output) {
+        _p = PredictorFactory.makePredictor(modelInfo._boosterBytes, null, false);
+        _di = di;
+        _sparse = output._sparse;
+
+        if (_p.getNumClass() > 2) {
+            throw new UnsupportedOperationException("Calculating contributions is currently not supported for multinomial models.");
+        }
+        GBTree gbTree = (GBTree) _p.getBooster();
+        RegTree[] trees = gbTree.getGroupedTrees()[0];
+        _nodeWeights = new double[trees.length][];
+        for (int i = 0; i < trees.length; i++) {
+            _nodeWeights[i] = new double[trees[i].getStats().length];
+        }
+    }
+
+    @Override
+    public void map(Chunk[] chks, NewChunk[] idx) {
+        MutableOneHotEncoderFVec inputVec = new MutableOneHotEncoderFVec(_di, _sparse);
+        double[] input = new double[chks.length - 1];
+        for (int row = 0; row < chks[0]._len; row++) {
+            double weight = chks[input.length].atd(row);
```

Comment:
```
double[] input = new double[chks.length - 1];
.
.
double weight = chks[input.length].atd(row);

look strange to me
```

Reply:
that would be too pretty :) ok, just kidding - will improve!

Reply:
done